wallet/wallet_v4.tact

```diff
@@ -0,0 +1,121 @@
+interface Plugin {
+    internal requested_funds#f06c7567(query_id: Uint64);
+    internal notify_removed#e4737472(query_id: Uint64);
+    internal notify_installed#6e6f7465(query_id: Uint64);
+    internal notify_removed#64737472(query_id: Uint64);
+}
+
+contract Wallet {
+    var seqno: Uint32
+    var subwallet: Uint32
+    var pubkey: Pubkey
+    var plugins: HashSet[Plugin]
+
+    bounced default() {
+        // ignore bounced messages
+    }
+
+    internal default() {
+        // ignore simple transfers
+    }
+
+    internal request_funds#706c7567(
+        query_id: Uint64,
+        requested_amount: Coins,
+        requested_extra: ExtraCurrencies,
+        msg.sender: Plugin,
+    ) {
+        verify(self.balance() - msg.amount >= requested_amount, "Not enough funds");
+
+        send(InboundCoins)
+        msg.sender.requested_funds(
+            query_id,
+            msg.amount: requested_amount,
+            msg.extra: requested_extra,
+        );
+    }
+
+    internal remove_plugin#64737472(query_id: Uint64, msg.sender: Plugin) {
+        self.remove_plugin_from_list(msg.sender);
+
+        if msg.bounceable {
+            send(InboundCoins) msg.sender.notify_removed(query_id);
+        }
+    }
+
+    external pre_condition(body: Slice) -> Slice {
+        let sign = body.read(Signature);
+        if sign.verify(body, self.pubkey) {
+            self.accept();
+            self.seqno += 1;
+            self.commit();
+            return body;
+        } else {
+            thrown("Invalid signature");
+        }
+    }
+
+    external simple_transfer#0(messages: List[(Cell, Sendmode)]) -> Slice {
```

Comment:
Option B: have a special support for cases where we tie the array to the refs, so we don't have an explicit length prefix at TL-B level.

We can invent something like `List[T,B]` where the first type is for the ref and the second one for the bits to read. 

The default is to read no bits: `List[T, B=()]`, so saying `List[Blah]` would parse each ref'd cell as `Blah` and read no bits.

```
List[RawMessage] // typed refs, nothing read from bits.
List[RawMessage, Sendmode] // typed refs, some type to read from bits.
```

Reply:
That'd be some weird list, but in the basic configuration w/o extra bits would look pretty innocent - you have some typed cells `List[Blah]` and everything feels quite normal. The extra type is a an extra feature to pack bits more densely.

Compiler should check what's the maximum possible number of cells you could fit this way (maybe 2 or 3 because you used up some refs in the definition already), and check that you have enough bits for the extra `B` type.